Update miniquad and imgui dependencies

laumann opened the pull request:
The EventHandlerFree trait is gone now, so we don't have an owned
Context. Instead implement the EventHandler trait, and leave it for
later to figure out how to support the clipboard.

The imgui demo still has clipboard support...

laumann committed:
Update miniquad and imgui dependencies

The EventHandlerFree trait is gone now, so we don't have an owned
Context. Instead implement the EventHandler trait, and leave it for
later to figure out how to support the clipboard.

Clipboard support is now provided via the clipboard crate, as that's how
the imgui-rs examples are doing it.

laumann commented:
@not-fl3 I think a few more changes could be in order

 - clean up some of the commented stuff
 - maybe remove the inner platform module? `set_ctx` isn't really used anymore
 - I've ruined the clipboard, not quite sure how to handle that

laumann commented:
Clipboard support is back!

not-fl3 commented on Cargo.toml:
```diff
@@ -9,9 +9,10 @@ repository = "https://github.com/not-fl3/imgui-miniquad-render"
 
 [dependencies]
-miniquad = { version = "0.3.0-alpha.20" }
+clipboard = "0.5"
```
I would rather use miniquad's `Context::clipboard_set`/`Context::clipboard_get`

not-fl3 replied:
oh, but ClipboardBackend really wants to be 'static :(

not-fl3 commented:
Hi @laumann :wave: 

> maybe remove the inner platform module? set_ctx isn't really used anymore

Yeah, getting a 'static context reference is hard now :( 
`request_quit` from `platform` is a bit better than `exit(0)` - it gives a nice `on_quit` message to gives app the opportunity to persist some state before closing the window. So it is worth preserving 

but some dirty hack like this will do, I guess:
```diff
diff --git a/src/lib.rs b/src/lib.rs
index 035ce66..648964d 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -34,21 +34,15 @@ impl ClipboardBackend for ClipboardSupport {
 
 /// Platform dependent APIs not directly connected to imgui
 pub mod platform {
-    use miniquad::Context as QuadContext;
-    use std::{cell::RefCell, rc::Rc};
+    use std::cell::RefCell;
 
-    static mut QUAD_CONTEXT: Option<Rc<RefCell<QuadContext>>> = None;
-
-    pub(crate) fn set_ctx(ctx: Rc<RefCell<QuadContext>>) {
-        unsafe { QUAD_CONTEXT = Some(ctx) };
+    thread_local! {
+        pub static QUIT_ORDERED: RefCell<bool> = RefCell::new(false);
     }
 
     /// Close window. "quit" event will be triggered.
     pub fn request_quit() {
-        unsafe { QUAD_CONTEXT.as_ref() }
-            .unwrap()
-            .borrow_mut()
-            .request_quit();
+        QUIT_ORDERED.with(|q| *q.borrow_mut() = true);
     }
 }
 
@@ -242,6 +236,9 @@ impl EventHandler for Stage {
     }
 
     fn draw(&mut self, ctx: &mut QuadContext) {
+        if platform::QUIT_ORDERED.with(|q| *q.borrow()) {
+            ctx.request_quit();
+        }
         let draw_data = {
             let io = self.imgui.io_mut();
             let now = std::time::Instant::now();
@@ -384,15 +381,8 @@ impl Window {
         }
     }
 
-    pub fn main_loop(self, on_draw: impl FnMut(&mut imgui::Ui) -> ()) -> ! {
-        let on_draw = Box::new(on_draw);
-
-        // Allocate `closure` on the heap and erase the lifetime bound.
-        // This is safe because we will never leave this function (alive)
-        // The same applies for closure in on_init
-        let closure: Box<dyn FnMut(&mut imgui::Ui)> = Box::new(on_draw);
-        let closure: Box<dyn FnMut(&mut imgui::Ui) + 'static> =
-            unsafe { std::mem::transmute(closure) };
+    pub fn main_loop(self, on_draw: impl FnMut(&mut imgui::Ui) -> () + 'static) -> ! {
+        let closure = Box::new(on_draw);
 
         miniquad::start(conf::Conf::default(), move |ctx| {
             Box::new(Stage::new(ctx, closure, self.on_init, self.on_quit))
```

laumann commented:
Hi @not-fl3 :smile: Alright, thanks, I'll redo this tomorrow.

laumann committed:
platform: Make request_quit() work again

request_quit() from platform is a bit better than exit(0) - it gives a
nice on_quit message to gives app the opportunity to persist some state
before closing the window.

Co-authored-by: Fedor Logachev <not-fl3@gmail.com>

laumann commented:
I applied the dirty-hack provided above - still not sure what to do about the clipboard.

not-fl3 commented on src/lib.rs:
```diff
@@ -378,18 +381,11 @@ impl Window {
-        let closure: Box<dyn FnMut(&mut imgui::Ui)> = Box::new(on_draw);
-        let closure: Box<dyn FnMut(&mut imgui::Ui) + 'static> =
-            unsafe { std::mem::transmute(closure) };
+    pub fn main_loop(self, on_draw: impl FnMut(&mut imgui::Ui) -> () + 'static) -> ! {
```
Oh, I feel that I should not add this `'static` to this diff (as much as I would like to..)

examples are not building with this `'static` :(

laumann committed:
Remove commented line

not-fl3 commented:
> I applied the dirty-hack provided above - still not sure what to do about the clipboard.

I have two solutions: 

- merge as is
- or move clipboard stuff into user `on_init` implementation, something like this:
```diff
diff --git a/Cargo.toml b/Cargo.toml
index 0899207..64b50f2 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -9,10 +9,12 @@ repository = "https://github.com/not-fl3/imgui-miniquad-render"
 description = "miniquad based minimal imgui application wrapper"
 
 [dependencies]
-clipboard = "0.5"
 miniquad = { version = "0.3.12" }
 glam = {version = "0.8", features = ["scalar-math"] }
 imgui = { version = "0.8.2" }
 
+[dev-dependencies]
+clipboard = "0.5"
+
 #[patch.crates-io]
 #miniquad = { path = '../miniquad' }
diff --git a/examples/text_editor.rs b/examples/text_editor.rs
index fdc41bc..a1c5574 100644
--- a/examples/text_editor.rs
+++ b/examples/text_editor.rs
@@ -9,6 +9,31 @@ fn main() {
             let style = imgui.style_mut();
             style.window_rounding = 0.;
             style.use_light_colors();
+
+            {
+                use clipboard::ClipboardProvider;
+
+                /// Clipboard support from the clipboard crate
+                struct ClipboardSupport(pub clipboard::ClipboardContext);
+                impl ClipboardBackend for ClipboardSupport {
+                    fn get(&mut self) -> Option<String> {
+                        self.0.get_contents().ok()
+                    }
+
+                    fn set(&mut self, text: &str) {
+                        let _ = self.0.set_contents(text.to_owned());
+                    }
+                }
+
+                if let Some(clip_backend) = clipboard::ClipboardContext::new()
+                    .ok()
+                    .map(ClipboardSupport)
+                {
+                    imgui.set_clipboard_backend(clip_backend);
+                } else {
+                    eprintln!("failed to initialize clipboard!");
+                }
+            }
         })
         .main_loop(|ui| {
             let [width, height] = ui.io().display_size;
diff --git a/src/lib.rs b/src/lib.rs
index 035ce66..859a8b7 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -135,15 +115,6 @@ impl Stage {
             )
         };
 
-        if let Some(clip_backend) = clipboard::ClipboardContext::new()
-            .ok()
-            .map(ClipboardSupport)
-        {
-            imgui.set_clipboard_backend(clip_backend);
-        } else {
-            eprintln!("failed to initialize clipboard!");
-        }
-
         //platform::set_ctx(ctx.clone());
 
         if let Some(on_init) = on_init {
```

The first solution solves the problem, but in a way I do not really like.
Second, do not solve the problem, the crate just does not support the clipboard at all, but there is a user-land workaround in the examples.

not-fl3 commented:
This is waaaay beyound the scope of this PR, but:

Actually, there is a real solution, I am just procrastinating on it since forever :/

We have exactly the same problem in macroquad - macroquad really wants a nice 'static reference to the context.
I want to move the entire Context to the `thread_local`. And stop passing this reference around all the time completely. 
It will not change anything at all - on most platforms Context depends on lots of little thread_locals anyway. 

Said all this, I believe that merging this PR with `clipboard` crate is OK, not much we can do about it right now :)
